File: app/src/main/java/me/zhanghai/android/files/provider/remote/ParcelableDirectoryStream.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote;

import android.os.Parcel;
import android.os.Parcelable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import java8.nio.file.DirectoryIteratorException;
import java8.nio.file.DirectoryStream;
import java8.nio.file.Path;
import java9.lang.Iterables;
import me.zhanghai.android.files.provider.common.PathListDirectoryStream;

public class ParcelableDirectoryStream implements Parcelable {

    @Nullable
    private final List<Path> mPaths;

    private transient boolean mGetCalled;

    public ParcelableDirectoryStream(@Nullable DirectoryStream<Path> directoryStream)
            throws IOException {
        mPaths = getPaths(directoryStream);
    }

    @Nullable
    private static List<Path> getPaths(@Nullable DirectoryStream<Path> directoryStream)
            throws IOException {
        if (directoryStream == null) {
            return null;
        }
        List<Path> paths = new ArrayList<>();
        try {
            Iterables.forEach(directoryStream, paths::add);
        } catch (DirectoryIteratorException e) {
            throw e.getCause();
        }
        return paths;
    }

    @Nullable
    public DirectoryStream<Path> get() {
        if (mGetCalled) {
            throw new IllegalStateException("Already called get() on this instance");
        }
        mGetCalled = true;
        return mPaths != null ? new PathListDirectoryStream(mPaths, path -> true) : null;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeList(mPaths);
    }

    protected ParcelableDirectoryStream(Parcel in) {
        //noinspection unchecked
        mPaths = in.readArrayList(Path.class.getClassLoader());
    }

    public static final Creator<ParcelableDirectoryStream> CREATOR =
            new Creator<ParcelableDirectoryStream>() {
                @Override
                public ParcelableDirectoryStream createFromParcel(Parcel source) {
                    return new ParcelableDirectoryStream(source);
                }
                @Override
                public ParcelableDirectoryStream[] newArray(int size) {
                    return new ParcelableDirectoryStream[size];
                }
            };
}

File: app/src/main/java/me/zhanghai/android/files/provider/remote/ParcelableDirectoryStream.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote

import android.os.Parcel
import android.os.Parcelable
import java8.nio.file.DirectoryIteratorException
import java8.nio.file.DirectoryStream
import java8.nio.file.Path
import me.zhanghai.android.files.provider.common.PathListDirectoryStream
import me.zhanghai.android.files.util.ParcelSlicedList
import java.io.IOException

class ParcelableDirectoryStream : Parcelable {
    private val paths: List<Path>

    val value: DirectoryStream<Path>
        get() = PathListDirectoryStream(paths) { true }

    @Throws(IOException::class)
    constructor(value: DirectoryStream<Path>) {
        paths = try {
            value.toList()
        } catch (e: DirectoryIteratorException) {
            throw e.cause!!
        }
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        @Suppress("UNCHECKED_CAST")
        dest.writeParcelable(ParcelSlicedList(paths as List<Parcelable>), flags)
    }

    private constructor(source: Parcel) {
        @Suppress("UNCHECKED_CAST")
        paths = source.readParcelable<ParcelSlicedList<Parcelable>>(Path::class.java.classLoader)!!
            .list as List<Path>
    }

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.Creator<ParcelableDirectoryStream> {
            override fun createFromParcel(source: Parcel): ParcelableDirectoryStream =
                ParcelableDirectoryStream(source)

            override fun newArray(size: Int): Array<ParcelableDirectoryStream?> = arrayOfNulls(size)
        }
    }
}
